### PR TITLE
 [PVR] Feature: Separate windows for "Timer Rules" and "Timers"

### PR DESCRIPTION
**Pull request by ksooo**
Currently, PVR timer window contains both "timer rules" (repeating timers) and "timers" (oneshot timers) in one list. Users complained (and I agree) that this unnecessarily clutters window content with actually different things. Thus, this PR introduces a new window for only the timer rules and leaves the timer window with only the "real" timers (like it was before Jarvis).

screenshot000
screenshot001
screenshot002

@Jalle19 @xhaggi others... What do you think?

### CODE REVIEW

**Comment by ksooo** [comment]
@ronie ping, for the confluence changes...

**Comment by ronie** [comment]
confluence code looks good to me.

**Comment by xhaggi** [comment]
why do we separate the window if this is only a possible type for a timer. would it be better to only have a filter which let you choose which kind of timer type you want to show?  (both, timer or rules)

**Comment by xhaggi** [comment]
btw instead of using a label which indicates the type why not introduce different icons infront of each timer entry. not sure why people complaining about that. i don't want to switch between two windows to check what are my next scheduled timers regardless of which type the timers are.

**Comment by ksooo** [comment]
> why do we separate the window if this is only a possible type for a timer.

Recording rules as timer types is purely technical stuff I "invented" because it was easy to implement with what we had already in our code base. But normal users do not understand and don't care about this, as several postings in the forum prove. For users, there are "recording rules" and "upcoming recordings". They treat this as completely different things. Users tell as that "all other" PVRs have two windows for that.

> would it be better to only have a filter which let you choose which kind of timer type you want to show? (both, timer or rules)

IMHO, no. Played with this as well and for me this felt not nearly as "natural" as two windows. Using the side blade to switch the view mode "persistently" is perfectly okay, but not if you just want to do a quick look at "the other" possible items in the window. 

> btw instead of using a label which indicates the type why not introduce different icons infront of each timer entry.

 Nice idea, but does not actually solve the problem the users complain about. Let's discuss/do this in another PR/thread, please.

> i don't want to switch between two windows to check what are my next scheduled timers regardless of which type the timers are. 

 Hehe. Many others want. ;-)

**Comment by metaron-uk** [comment]
Definitely a +1 from me. Being able to see at a glance what is going to record on what channel this evening without being swamped by all the outstanding 'rules' I've got set up is a godsend (talking of which, why isn't this evening's episode of Strictly recording...).

Having tested it out with yesterday's Millhouse build, I have some suggestions (maybe in this PR, maybe for subsequent PRs...):

1.  This needs a new context menu in the 'Timers' window - 'Edit Rule' or some such name. It would be good to have this in the 'Search' results and EPG context menus as well.
In this case 'Edit' should probably become 'Edit Timer'. (Confusing if you read the code, but makes sense on the UI imo)

2.  The Recording rules sideblade needs sort by 'Type' as an option

3.   It would be brilliant if a 'Timer Count' column could be added to Recording rules view to show which rules are 'busy' (have active 'timers') and which are 'stale' (no active timers). Maybe this could be merged with the 'Status' column e.g.: Disabled/Enabled/5 Timers?

4. If that is added, a sort by 'Status' sideblade option starts to make sense as well....

**Comment by ksooo** [comment]
@metaron-uk I like your suggestions. Feel free to come up with PRs. Don't want to extend this PR, though.

**Comment by xhaggi** [comment]
please wait until next week before merge it in. I'll take a deeper look at it tomorrow or within the next days.

**Comment by ksooo** [comment]
@xhaggi sure. All fine. Always listening carefully to your comments. :-)

**Comment by Jalle19** [comment]
Haven't looked at the code yet but I support the change. tvheadend itself also separates recording rules from actual upcoming recordings.

**Comment by Jalle19** [comment]
Would it make sense to split the window into a real separate window (maybe not with this PR but in the future? Remember we used to have all PVR windows as a single pseudo-window and it was a pain in the ass for @xhaggi to split them, but the end result was 100% worth it.

**Comment by xhaggi** [comment]
@Jalle19 yep that's the first I was thinking about, but did not talked to @ksooo yet. We should not introduce such a hack to re-use window code. We should derive from a common abstract class where we put in all the dupe-code we need.

**Comment by ksooo** [comment]
No problems with introducing a common (abstract - don't know this detail yet) base class CGUIWindowPVRTimersBase and derive CGUIWindowPVRTimers and CGUIWindowPVRRecordingRules from this. Makes sense, altough the derived classes would currently contain VERY little code.

Shall I also introduce MyPVRRecordingRules.xml which (for now) would be a 99% copy of MyPVRTimers.xml?

**Comment by Jalle19** [comment]
The skin part I'm not sure abouy, @ronie is it possible to use an include somehow?

**Comment by ronie** [comment]
please don't add new xml files to the skin, unless strictly needed.
we're trying to reduce the number of xml files for skins, not expand it.

both window should be able to use the same xml i think?
skins can include (or show/hide) certain parts by checking which of the two windows is visible.

**Comment by ksooo** [comment]
Currently I'm using the same xml file for both Windows. Works just fine. So, I propose to stick with the one xml approach.

**Comment by ksooo** [comment]
@xhaggi refactored. better now?

**Comment by ksooo** [comment]
Changed internal and external (UI) wording from "recording rules" to "timer rules", as requested internally with @xhaggi

**Comment by xhaggi** [comment]
@ksooo thanks :+1:

**Comment by xhaggi** [comment]
you need an osx sync?

**Comment by ksooo** [comment]
> you need an osx sync?

Will provide that myself, thanks. But VS solution files update would be very welcome (don't have Windows).

**Comment by xhaggi** [comment]
me too, i don't have win ;)

**Comment by ksooo** [comment]
@Razzeee mind to provide the VS sln file update needed due to introduction of new files in this PR?

**Comment by Razzeee** [comment]
Pick https://github.com/Razzeee/xbmc/commit/d7cc978ac0b70010c30b3fdd2cb04c80ae44f87c
Compiles fine for me, but I'm having problems running it, likely only my system as this one has a new installation of VS and it might be borked.

File: Kodi.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -226,8 +228,8 @@
 		39BD2AD91B845D40004A5A15 /* DialogHelper.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 39BD2AD71B845D40004A5A15 /* DialogHelper.cpp */; };
 		39C38CCA1BBFF1EE000F59F5 /* InputCodingTableKorean.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 39C38CC81BBFF1EE000F59F5 /* InputCodingTableKorean.cpp */; };
 		39C38CCB1BBFF1EE000F59F5 /* InputCodingTableKorean.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 39C38CC81BBFF1EE000F59F5 /* InputCodingTableKorean.cpp */; };
-		39C38CE11BCD600E000F59F5 /* FFmpegImage.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 39C38CDF1BCD600E000F59F5 /* FFmpegImage.cpp */; settings = {ASSET_TAGS = (); }; };
```
Comment (Jalle19) [inline Kodi.xcodeproj/project.pbxproj]: Seems unrelated?

File: xbmc/pvr/windows/GUIWindowPVRGuide.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ namespace PVR
 
   protected:
     void UpdateSelectedItemPath();
+    virtual std::string GetDirectoryPath(void) { return ""; }
```
Comment (Jalle19) [inline xbmc/pvr/windows/GUIWindowPVRGuide.h]: Could you add `override` since the parent is a pure virtual? Same in the other derived class.

File: Kodi.xcodeproj/project.pbxproj (hunk 1 above)
Comment (ksooo) [inline Kodi.xcodeproj/project.pbxproj]: Come on! This gets auto generated by Xcode and is obviously not harmful. ;-)

**Comment by ksooo** [comment]
@Razzeee thanks

File: xbmc/pvr/windows/GUIWindowPVRGuide.h (hunk 2 above)
Comment (ksooo) [inline xbmc/pvr/windows/GUIWindowPVRGuide.h]: Sure. Will do.

**Comment by ksooo** [comment]
@Jalle19 now with the override's you suggested.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
Build breaker on IOS, related to this PR! Have to check the xcode project update, I guess. Think, I forgot to add the new files for IOS.

**Comment by ksooo** [comment]
jenkins build this please

File: Kodi.xcodeproj/project.pbxproj (hunk 1 above)
Comment (Jalle19) [inline Kodi.xcodeproj/project.pbxproj]: Okay, in that case

**Comment by Jalle19** [comment]
Seems okay to me

**Comment by ksooo** [comment]
Okay then, if nobody objects I will merge this tomorrow.
